pkg/networkpolicy/packet_test.go

```diff
@@ -0,0 +1,231 @@
+package networkpolicy
+
+import (
+	"net"
+	"testing"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+// captureChunk is the number of bytes transfered to user-space
+// in the nfqueue
+const captureChunk = 128
+
+func TestUdpFragmentIPv6(t *testing.T) {
+	// The test-data contains 3 packets:
+	// 1. First fragment of an UDP packet
+	// 2. Fragment containing the rest of the UDP packet
+	// 3. An unfragmented reply UDP packet
+	// We expect:
+	// 1. Identified as an UDP packet
+	// 2. Unknown protocol (44)
+	// 3. Identified as an UDP packet
+	tests := []struct {
+		name     string
+		input    []byte
+		err      bool
+		expected packet
+	}{
+		{
+			name:  "UDP first fragment",
+			input: packetsUdpFragIPv6[0][14:(14 + captureChunk)],
+			expected: packet{
+				family:  v1.IPv6Protocol,
+				proto:   v1.ProtocolUDP,
+				dstIP:   net.ParseIP("fd00::c0a8:101"),
+				dstPort: 5001,
+			},
+		},
+		{
+			name:  "UDP not-first fragment",
+			input: packetsUdpFragIPv6[1][14:(14 + captureChunk)],
+			err:   true,
+		},
+		{
+			name:  "UDP packet (un-fragmented)",
+			input: packetsUdpFragIPv6[2][14:],
+			expected: packet{
+				family:  v1.IPv6Protocol,
+				proto:   v1.ProtocolUDP,
+				srcIP:   net.ParseIP("fd00::c0a8:101"),
+				srcPort: 5001,
+			},
+		},
+	}
+	for _, tc := range tests {
+		packet, err := parsePacket(tc.input)
+		if err != nil {
+			if !tc.err {
+				t.Fatalf("%s: unexpected error: %v", tc.name, err)
+			}
+			continue
+		}
+		comparePacket(t, tc.name, tc.expected, packet)
+	}
+}
+
+func comparePacket(t *testing.T, tc string, expected, got packet) {
+	if got.family != expected.family {
+		t.Fatalf("%s: family: expected=%v, got=%v", tc, expected.family, got.family)
+	}
+	if got.proto != expected.proto {
+		t.Fatalf("%s: proto: expected=%v, got=%v", tc, expected.proto, got.proto)
+	}
+	// Compare other fields only if expected (never compare id and payload)
+	if expected.srcIP != nil {
+		if !got.srcIP.Equal(expected.srcIP) {
+			t.Fatalf("%s: srcIP: expected=%v, got=%v", tc, expected.srcIP, got.srcIP)
+		}
+	}
+	if expected.srcPort > 0 {
+		if got.srcPort != expected.srcPort {
+			t.Fatalf("%s: srcPort: expected=%v, got=%v", tc, expected.srcPort, got.srcPort)
+		}
+	}
+	if expected.dstIP != nil {
+		if !got.dstIP.Equal(expected.dstIP) {
+			t.Fatalf("%s: dstIP: expected=%v, got=%v", tc, expected.dstIP, got.dstIP)
+		}
+	}
+	if expected.dstPort > 0 {
+		if got.dstPort != expected.dstPort {
+			t.Fatalf("%s: dstPort: expected=%v, got=%v", tc, expected.dstPort, got.dstPort)
+		}
+	}
+}
+
+var packetsUdpFragIPv6 = [][]byte{
```

Comment:
seems like this would make more sense as three separate variables with useful names.
also, maybe split into L2 header, L3 header, payload, with some minimal comments?

Reply:
I saved the original pcap files in `test-data/` so people can analyze it with `wireshark` for instance